src/Line.cpp

```diff
@@ -1,33 +1,88 @@
-//
-// Created by rolf on 18-4-19.
-//
-
 #include "../include/roboteam_utils/Line.h"
 #include "../include/roboteam_utils/LineSegment.h"
 
 namespace rtt {
+Line::Line(const Vector2 &location, const Vector2 &direction) {
+    this->location = location;
+    this->direction = direction;
+    if (location == direction) {
+        std::cout << "Warning: you created an undefined line, because location == direction. Note that Lines have an infinite length. If you want to have a Line with finite length"
+                     " then use the LineSegment class instead." << std::endl;
+    }
+}
+
 Line::Line(const LineSegment &other) noexcept {
-    start=other.start;
-    end=other.end;
+    location = other.start;
+    direction = other.end;
+    if (location == direction) {
+        std::cout << "Warning: you created an undefined line, because location == direction. Note that Lines have an infinite length. If you want to have a Line with finite length"
+                     " then use the LineSegment class instead." << std::endl;
+    }
 }
-double Line::length() const {
-    return (end - start).length();
+
+double Line::distanceToLine(const Vector2 &point) const {
+    return (this->project(point) - point).length();
 }
 
-double Line::length2() const {
-    return (end - start).length2();
+Vector2 Line::project(const Vector2 &point) const {
+    Vector2 AB = direction - location;
+    Vector2 AP = point - location;
+    return location + AB * (AP.dot(AB) / AB.length2());
 }
 
-double Line::slope() const {
-    return (end.y - start.y)/(end.x - start.x);
+std::optional<Vector2> Line::intersect(const Line &line) const {
+    return intersect(location, direction, line.location, line.direction);
 }
 
-bool Line::isVertical() const {
-    return (end.x == start.x) && (end.y != start.y);
+std::optional<Vector2> Line::intersect(const Vector2 line1Location, const Vector2 line1Direction, const Vector2 line2Location, const Vector2 line2Direction) {
+    Vector2 A = line1Location - line1Direction;
+    Vector2 B = line2Location - line2Direction;
+    double denom = A.cross(B);
+    if (denom != 0) {
+        Vector2 C = line1Location - line2Location;
+        double numer = C.cross(A);
+        double u = numer / denom;
+        return line2Location - B * u;
+    }
+    return std::nullopt;
 }
```

Comment:
I explained this in the documentation, when two lines are parallel (which includes colinear and overlapping) I don't return any point (because there are either infinitely many intersections or none at all), because of performance reasons. This also happened in the previous implementation.

Reply:
That is indeed the behaviour according to the documentation.